### DISCUSSION TITLE
Does Lit automatically remove declarative event listers?

### DISCUSSION THREAD

**Discussion by minicatsCB** [lit/lit#4781] [Q&A] [answered]
Hi! I am reading the documentation about adding event listeners in Lit. I am aware that if I do it programatically, I should remove them in the `disconnectedCallback()` hook. But what if I do it declaratively in the HTML template? I can't see any mention to that in the [documentation](https://lit.dev/docs/v2/components/events/#adding-event-listeners-in-the-element-template).

If necessary, I can create a PR to include that. However, I’d like to confirm with you first. 😊

**Comment by justinfagnani** [comment]
Listeners aren't but removed, but they didn't need to be. They will be collected my the garbage collector.

**Reply by minicatsCB** [reply]
Sorry, I didn't understand the first part. Are they removed automatically by Lit or not?

**Reply by tbroyer** [reply, marked as answer]
Lit won't remove them, because there's no need to do that, they'll be garbage-collected at the same time as the element they're attached to.

If you do add listeners programmatically, you only need to remove them in disconnectedCallback if you added them to elements not in your control (window, document, children in the light DOM, any other element in the document, etc.)

**Reply by justinfagnani** [reply]
Sorry, that was a swypo. I meant "aren't". Will edit.

**Reply by justinfagnani** [reply]
To clarify @tbroyer's comment here:

> you only need to remove them in disconnectedCallback if you added them to elements not in your control

Browser GCs can collect cycles. So a host element that references a child that references back to the host via an event listener is fine, as long as the the host and child element don't have any external references to them that _otherwise_ prevent them from being GC'ed.

With shadow DOM, there's almost never an external reference to a shadow child, so you don't have to worry about cycles between shadow children and the host.

But if you add an event listener to an _external_ object like `window`, then `window` is now holding a reference via that event listener's closure, back to your host and will prevent it from being GC'ed. Those are the event listeners you must clean up.

**Reply by minicatsCB** [reply]
Interesting!

@justinfagnani what do you exactly mean with "cycles between shadow children and the host"? Do you refer to web components lifecycle?

**Reply by justinfagnani** [reply]
Reference cycles. There are references from the host to the shadow children via the shadow root, then the children of the shadow root, etc. There can be a reference from the child to the host via an event listener on the child that references the host (often via `this`).
